server/project/project.go

```diff
@@ -228,6 +228,18 @@ func (s *Server) Update(ctx context.Context, q *ProjectUpdateRequest) (*v1alpha1
 		return nil, err
 	}
 
+	for _, cluster := range difference(q.Project.Spec.DestinationClusters(), oldProj.Spec.DestinationClusters()) {
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceClusters, rbacpolicy.ActionUpdate, cluster); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, repoUrl := range difference(q.Project.Spec.SourceRepos, oldProj.Spec.SourceRepos) {
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceRepositories, rbacpolicy.ActionUpdate, repoUrl); err != nil {
+			return nil, err
+		}
+	}
+
```

Comment:
Theres a missing check that needs to be done, which is: If the user is modifying the clusterResourceWhitelist, or the namespaceResourceBlacklist, they need update permissions on every cluster in the project list.

We basically need to allow a developer role, from escalating the sandbox that a operator is giving that developer, while still giving the developer freedom to manage their project tokens, project roles, repos, etc...

Reply:
let me get this right, if they make **any** change to either of those lists, then we need the check all the clusters?

Reply:
```golang
	clusterResourceWhitelistsEqual := reflect.DeepEqual(q.Project.Spec.ClusterResourceWhitelist, oldProj.Spec.ClusterResourceWhitelist)
	namespacesResourceBlacklistsEqual := reflect.DeepEqual(q.Project.Spec.NamespaceResourceBlacklist, oldProj.Spec.NamespaceResourceBlacklist)
	if !clusterResourceWhitelistsEqual || !namespacesResourceBlacklistsEqual {
		for _, cluster := range q.Project.Spec.DestinationClusters() {
			if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceClusters, rbacpolicy.ActionUpdate, cluster); err != nil {
				return nil, err
			}
		}
	}
```

Reply:
amirite?